File: cpp/theme.cpp
#include "theme.h"

#include <QDir>
#include <QFile>
#include <QFileInfo>
#include <QRegularExpression>
#include <QTextStream>

OmarchyTheme::OmarchyTheme(QObject *parent)
    : QObject(parent)
{
    m_themeDir = QDir::homePath() + QStringLiteral("/.local/state/omarchy/current/theme");
    m_colorsFile = m_themeDir + QStringLiteral("/colors.toml");
    m_currentDir = QFileInfo(QDir(m_themeDir).absolutePath()).absolutePath(); // .../current

    loadTheme();

    connect(&m_watcher, &QFileSystemWatcher::fileChanged, this, [this](const QString &) {
        loadTheme();
        setupWatcher();
    });
    connect(&m_watcher, &QFileSystemWatcher::directoryChanged, this, [this](const QString &) {
        loadTheme();
        setupWatcher();
    });
    setupWatcher();
}

void OmarchyTheme::setupWatcher()
{
    const QStringList watchedFiles = m_watcher.files();
    const QStringList watchedDirs = m_watcher.directories();
    for (const QString &path : {m_currentDir, m_themeDir, m_colorsFile}) {
        if (QFileInfo::exists(path)
            && !watchedFiles.contains(path)
            && !watchedDirs.contains(path)) {
            m_watcher.addPath(path);
        }
    }
}

static QString parseTomlString(const QString &content, const QString &key, const QString &fallback)
{
    // Minimal flat TOML: key = "value" or key = 'value'
    const QRegularExpression re(
        QStringLiteral("^\\s*%1\\s*=\\s*[\"']([^\"']*)[\"']").arg(QRegularExpression::escape(key)),
        QRegularExpression::MultilineOption);
    const auto m = re.match(content);
    if (m.hasMatch())
        return m.captured(1);
    return fallback;
}

void OmarchyTheme::loadTheme()
{
    QFile f(m_colorsFile);
    if (!f.open(QIODevice::ReadOnly | QIODevice::Text))
        return;

    const QString content = QString::fromUtf8(f.readAll());
    m_mode = parseTomlString(content, QStringLiteral("mode"), m_mode);
    m_accent = parseTomlString(content, QStringLiteral("accent"), m_accent);
    m_background = parseTomlString(content, QStringLiteral("background"), m_background);
    m_darkBackground = parseTomlString(content, QStringLiteral("dark_background"), m_darkBackground);
    m_darkerBackground = parseTomlString(content, QStringLiteral("darker_background"), m_darkerBackground);
    m_lighterBackground = parseTomlString(content, QStringLiteral("lighter_background"), m_lighterBackground);
    m_foreground = parseTomlString(content, QStringLiteral("foreground"), m_foreground);
    m_darkForeground = parseTomlString(content, QStringLiteral("dark_foreground"), m_darkForeground);
    m_lightForeground = parseTomlString(content, QStringLiteral("light_foreground"), m_lightForeground);
    m_muted = parseTomlString(content, QStringLiteral("muted"), m_muted);
    m_selection = parseTomlString(content, QStringLiteral("selection"), m_selection);
    m_red = parseTomlString(content, QStringLiteral("red"), m_red);
    m_green = parseTomlString(content, QStringLiteral("green"), m_green);
    m_yellow = parseTomlString(content, QStringLiteral("yellow"), m_yellow);
    m_orange = parseTomlString(content, QStringLiteral("orange"), m_orange);
    m_blue = parseTomlString(content, QStringLiteral("blue"), m_blue);
    m_cyan = parseTomlString(content, QStringLiteral("cyan"), m_cyan);
    m_magenta = parseTomlString(content, QStringLiteral("magenta"), m_magenta);

    emit themeChanged();
}

File: src/theme.py
"""
Theme manager for Omarchy.
Parses ~/.local/state/omarchy/current/theme/colors.toml
and monitors it using QFileSystemWatcher to support real-time hot-reloading.
"""

import os
from pathlib import Path
import tomllib
from PySide6.QtCore import QObject, Signal, Property, QFileSystemWatcher, Slot
from PySide6.QtGui import QColor


class OmarchyTheme(QObject):
    themeChanged = Signal()

    def __init__(self, parent=None):
        super().__init__(parent)

        self._theme_dir = Path.home() / ".local" / "state" / "omarchy" / "current" / "theme"
        self._colors_file = self._theme_dir / "colors.toml"

        # Defaults (Nord-like / Omarchy dark fallback)
        self._mode = "dark"
        self._accent = "#81a1c1"
        self._background = "#2e3440"
        self._dark_background = "#222730"
        self._darker_background = "#191c23"
        self._lighter_background = "#3b4252"
        self._foreground = "#d8dee9"
        self._dark_foreground = "#667080"
        self._light_foreground = "#adb5c4"
        self._muted = "#4c566a"
        self._selection = "#434c5e"
        self._red = "#bf616a"
        self._green = "#a3be8c"
        self._yellow = "#ebcb8b"
        self._orange = "#d5967a"
        self._blue = "#81a1c1"
        self._cyan = "#88c0d0"
        self._magenta = "#b48ead"

        self.load_theme()

        # Set up watcher
        self._watcher = QFileSystemWatcher(self)
        self._setup_watcher()

    def _setup_watcher(self):
        paths_to_watch = []
        if self._colors_file.exists():
            paths_to_watch.append(str(self._colors_file))
        if self._theme_dir.exists():
            paths_to_watch.append(str(self._theme_dir))

        if paths_to_watch:
            self._watcher.addPaths(paths_to_watch)
            self._watcher.fileChanged.connect(self._on_file_changed)
            self._watcher.directoryChanged.connect(self._on_dir_changed)

    @Slot(str)
    def _on_file_changed(self, path):
        self.load_theme()
        # Re-add file if it was replaced/atomic-renamed
        if self._colors_file.exists() and str(self._colors_file) not in self._watcher.files():
            self._watcher.addPath(str(self._colors_file))

    @Slot(str)
    def _on_dir_changed(self, path):
        self.load_theme()
        if self._colors_file.exists() and str(self._colors_file) not in self._watcher.files():
            self._watcher.addPath(str(self._colors_file))

    def load_theme(self):
        if not self._colors_file.is_file():
            return

        try:
            with open(self._colors_file, "rb") as f:
                data = tomllib.load(f)

            self._mode = data.get("mode", self._mode)
            self._accent = data.get("accent", self._accent)
            self._background = data.get("background", self._background)
            self._dark_background = data.get("dark_background", self._dark_background)
            self._darker_background = data.get("darker_background", self._darker_background)
            self._lighter_background = data.get("lighter_background", self._lighter_background)
            self._foreground = data.get("foreground", self._foreground)
            self._dark_foreground = data.get("dark_foreground", self._dark_foreground)
            self._light_foreground = data.get("light_foreground", self._light_foreground)
            self._muted = data.get("muted", self._muted)
            self._selection = data.get("selection", self._selection)
            self._red = data.get("red", self._red)
            self._green = data.get("green", self._green)
            self._yellow = data.get("yellow", self._yellow)
            self._orange = data.get("orange", self._orange)
            self._blue = data.get("blue", self._blue)
            self._cyan = data.get("cyan", self._cyan)
            self._magenta = data.get("magenta", self._magenta)

            self.themeChanged.emit()
        except Exception as e:
            print(f"Error reading Omarchy theme: {e}")

    # Properties exposed to QML
    @Property(str, notify=themeChanged)
    def mode(self) -> str:
        return self._mode

    @Property(QColor, notify=themeChanged)
    def accent(self) -> QColor:
        return QColor(self._accent)

    @Property(QColor, notify=themeChanged)
    def background(self) -> QColor:
        return QColor(self._background)

    @Property(QColor, notify=themeChanged)
    def darkBackground(self) -> QColor:
        return QColor(self._dark_background)

    @Property(QColor, notify=themeChanged)
    def darkerBackground(self) -> QColor:
        return QColor(self._darker_background)

    @Property(QColor, notify=themeChanged)
    def lighterBackground(self) -> QColor:
        return QColor(self._lighter_background)

    @Property(QColor, notify=themeChanged)
    def foreground(self) -> QColor:
        return QColor(self._foreground)

    @Property(QColor, notify=themeChanged)
    def darkForeground(self) -> QColor:
        return QColor(self._dark_foreground)

    @Property(QColor, notify=themeChanged)
    def lightForeground(self) -> QColor:
        return QColor(self._light_foreground)

    @Property(QColor, notify=themeChanged)
    def muted(self) -> QColor:
        return QColor(self._muted)

    @Property(QColor, notify=themeChanged)
    def selection(self) -> QColor:
        return QColor(self._selection)

    @Property(QColor, notify=themeChanged)
    def red(self) -> QColor:
        return QColor(self._red)

    @Property(QColor, notify=themeChanged)
    def green(self) -> QColor:
        return QColor(self._green)

    @Property(QColor, notify=themeChanged)
    def yellow(self) -> QColor:
        return QColor(self._yellow)

    @Property(QColor, notify=themeChanged)
    def orange(self) -> QColor:
        return QColor(self._orange)

    @Property(QColor, notify=themeChanged)
    def blue(self) -> QColor:
        return QColor(self._blue)

    @Property(QColor, notify=themeChanged)
    def cyan(self) -> QColor:
        return QColor(self._cyan)

    @Property(QColor, notify=themeChanged)
    def magenta(self) -> QColor:
        return QColor(self._magenta)
